module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php

```diff
@@ -0,0 +1,145 @@
+    ];
+
+    public $view;
+
+    public function __construct()
+    {
+        $this->view = new ViewModel();
+    }
+
+    public function indexAction()
+    {
+        $this->acceptableviewmodelselector($this->acceptCriteria);
+        $form = new AjaxImageUploadForm();
+        $this->view->form = $form;
+
+        if ($this->getRequest()->isPost()) {
+            // $form->setInputFilter($content->getInputFilter());
+            $data = array_merge_recursive(
+                $this->getRequest()->getPost()->toArray(),
+                $this->getRequest()->getFiles()->toArray()
+            );
+
+            $form->setData($data);
+            if (!empty($data["imageUpload"]) && $this->getRequest()->isXmlHttpRequest()) {
+                return $this->uploadImages();
+            }
+        }
+        return $this->view;
+    }
+
+    /**
+     * Get all files from all folders and list them in the gallery
+     */
+    public function filesAction()
+    {
+        $this->view->setTerminal(true);
+        $dir = new \RecursiveDirectoryIterator('public/userfiles/', \FilesystemIterator::SKIP_DOTS);
+        $it  = new \RecursiveIteratorIterator($dir, \RecursiveIteratorIterator::SELF_FIRST);
+        $it->setMaxDepth(3);
+        $files = [];
+        $i = 0;
+        foreach ($it as $file) {
+            if ($file->isFile()) {
+                $filepath = explode("public", $file->getRealPath()); // ugly workaround :(
+                $files[$i]["link"] = Json::encode($filepath[1]);
+                $files[$i]["filename"] = Json::encode($file->getFilename());
+                $i++;
+            }
+        }
+        return new JsonModel($files);
+    }
+
+    /**
+     * Create directories
+     * @return  void
+     */
+    private function createDirectories()
+    {
+        if (!is_dir('public/userfiles/')) {
+            mkdir('public/userfiles/');
+        }
+        if (!is_dir('public/userfiles/'.date("Y_M"))) {
+            mkdir('public/userfiles/'.date("Y_M"));
+        }
+        if (!is_dir('public/userfiles/'.date("Y_M").'/images/')) {
+            mkdir('public/userfiles/'.date("Y_M").'/images/');
+        }
+    }
+
+    /**
+     * Upload all images async
+     *
+     * @param  array $files
+     * @return Response containing headers with information about each image
+     */
+    private function uploadImages()
+    {
+        $this->view->setTerminal(true);
+        $adapter = new Http();
+        $size = new Size(['min'=>'10kB', 'max'=>'5MB','useByteString' => true]);
+        $isImage = new IsImage();
+        $extension = new Extension(['jpg','gif','png','jpeg','bmp','webp','svg'], true);
+
+        $this->createDirectories();
+        $adapter->setValidators([$size, $isImage, $extension]);
+        $adapter->setDestination('public/userfiles/'.date("Y_M").'/images/');
+        $this->upload($adapter);
+    }
+
+    /**
+     * @param  Zend\File\Transfer\Adapter\Http $adapter
+     */
+    private function upload($adapter)

```

Comment:
I think we can replace usage of <code>Zend\File\Transfer\Adapter\Http</code>, with define filters and validators in form::getInputFilterSpecification()